File: ControlLab/src/main/java/org/marsroboticsassociation/controllab/arm/ArmMetrics.java
package org.marsroboticsassociation.controllab.arm;

/**
 * Rolling, backlash-focused metrics for the Arm tab, computed against the load ground truth. Fed one
 * sample per control tick plus a notification on each target change (which resets the per-move
 * latches). All public angles are reported in degrees.
 *
 * <ul>
 *   <li><b>Steady-state error</b> — instantaneous |target − load|.
 *   <li><b>Overshoot</b> — max excursion of the load past the target in the direction of the move.
 *   <li><b>Settling time</b> — time from the last target change until the load first entered the
 *       position + velocity tolerance band.
 *   <li><b>Lost motion on reversal</b> — the lash gap |motor − load| the load lags across, tracked
 *       as the peak since the last commanded-direction reversal.
 *   <li><b>% time engaged</b> — an EMA of the plant's {@code isEngaged} flag.
 * </ul>
 */
public class ArmMetrics {

    private static final double SETTLE_POS_TOL_RAD = Math.toRadians(2.0);
    private static final double SETTLE_VEL_TOL_RAD = Math.toRadians(5.0);
    private static final double ENGAGED_EMA_ALPHA = 0.02; // ~ last few seconds at 60 Hz

    // Per-move latches
    private double moveStartTime = 0.0;
    private double moveDir = 0.0;          // sign(target - load) at last target change
    private double overshootDeg = 0.0;
    private double settleTimeSec = Double.NaN;
    private double lostMotionDeg = 0.0;    // peak lash gap since the last reversal

    // Live values
    private double targetRad = 0.0;
    private double lastErrorDeg = 0.0;
    private double lashGapDeg = 0.0;
    private double engagedEma = 1.0;
    private boolean sawTarget = false;

    /** Reset the per-move latches for a new commanded target. */
    public void onTargetChanged(double newTargetRad, double loadRad, double tSec) {
        double newDir = Math.signum(newTargetRad - loadRad);
        // A reversal is a direction flip relative to the previous move; reset the lash-gap peak.
        if (sawTarget && newDir != 0 && moveDir != 0 && newDir != moveDir) {
            lostMotionDeg = 0.0;
        }
        moveDir = newDir;
        targetRad = newTargetRad;
        moveStartTime = tSec;
        overshootDeg = 0.0;
        settleTimeSec = Double.NaN;
        sawTarget = true;
    }

    /** Feed one control-tick sample. */
    public void update(double tSec, double loadRad, double loadVelRad,
                       double motorRad, boolean engaged) {
        double errorRad = targetRad - loadRad;
        lastErrorDeg = Math.abs(Math.toDegrees(errorRad));
        lashGapDeg = Math.abs(Math.toDegrees(motorRad - loadRad));

        // Overshoot: how far past the target, in the direction of travel.
        if (moveDir != 0) {
            double past = Math.toDegrees((loadRad - targetRad) * moveDir);
            if (past > overshootDeg) overshootDeg = past;
        }

        // Settling time: first entry into the tolerance band after the target change.
        if (Double.isNaN(settleTimeSec)
                && Math.abs(errorRad) < SETTLE_POS_TOL_RAD
                && Math.abs(loadVelRad) < SETTLE_VEL_TOL_RAD) {
            settleTimeSec = tSec - moveStartTime;
        }

        // Lost motion / lash gap: peak since the last reversal.
        if (lashGapDeg > lostMotionDeg) lostMotionDeg = lashGapDeg;

        engagedEma += ENGAGED_EMA_ALPHA * ((engaged ? 1.0 : 0.0) - engagedEma);
    }

    /** Clear all state (used on engine reset). */
    public void reset() {
        moveStartTime = 0.0;
        moveDir = 0.0;
        overshootDeg = 0.0;
        settleTimeSec = Double.NaN;
        lostMotionDeg = 0.0;
        targetRad = 0.0;
        lastErrorDeg = 0.0;
        lashGapDeg = 0.0;
        engagedEma = 1.0;
        sawTarget = false;
    }

    public double steadyStateErrorDeg() { return lastErrorDeg; }
    public double overshootDeg()        { return overshootDeg; }
    public double settleTimeSec()       { return settleTimeSec; }
    public double lashGapDeg()          { return lashGapDeg; }
    public double lostMotionDeg()       { return lostMotionDeg; }
    public double pctEngaged()          { return engagedEma * 100.0; }
}

File: ControlLab/src/main/kotlin/org/marsroboticsassociation/controllab/arm/ArmMetrics.kt
package org.marsroboticsassociation.controllab.arm

import kotlin.math.abs
import kotlin.math.sign

/**
 * Rolling, backlash-focused metrics for the Arm tab, computed against the load ground truth. Fed
 * one sample per control tick plus a notification on each target change (which resets the per-move
 * latches). All public angles are reported in degrees.
 *
 * - **Steady-state error** — instantaneous |target − load|.
 * - **Overshoot** — max excursion of the load past the target in the direction of the move.
 * - **Settling time** — time from the last target change until the load first entered the
 *   position + velocity tolerance band.
 * - **Lost motion on reversal** — the lash gap |motor − load| the load lags across, tracked as the
 *   peak since the last commanded-direction reversal.
 * - **% time engaged** — an EMA of the plant's `isEngaged` flag.
 */
class ArmMetrics {
    // Per-move latches
    private var moveStartTime = 0.0
    private var moveDir = 0.0 // sign(target - load) at last target change
    private var overshootDegField = 0.0
    private var settleTimeSecField = Double.NaN
    private var lostMotionDegField = 0.0 // peak lash gap since the last reversal

    // Live values
    private var targetRad = 0.0
    private var lastErrorDeg = 0.0
    private var lashGapDegField = 0.0
    private var engagedEma = 1.0
    private var sawTarget = false

    /** Reset the per-move latches for a new commanded target. */
    fun onTargetChanged(newTargetRad: Double, loadRad: Double, tSec: Double) {
        val newDir = sign(newTargetRad - loadRad)
        // A reversal is a direction flip relative to the previous move; reset the lash-gap peak.
        if (sawTarget && newDir != 0.0 && moveDir != 0.0 && newDir != moveDir) {
            lostMotionDegField = 0.0
        }
        moveDir = newDir
        targetRad = newTargetRad
        moveStartTime = tSec
        overshootDegField = 0.0
        settleTimeSecField = Double.NaN
        sawTarget = true
    }

    /** Feed one control-tick sample. */
    fun update(
        tSec: Double,
        loadRad: Double,
        loadVelRad: Double,
        motorRad: Double,
        engaged: Boolean,
    ) {
        val errorRad = targetRad - loadRad
        lastErrorDeg = abs(Math.toDegrees(errorRad))
        lashGapDegField = abs(Math.toDegrees(motorRad - loadRad))

        // Overshoot: how far past the target, in the direction of travel.
        if (moveDir != 0.0) {
            val past = Math.toDegrees((loadRad - targetRad) * moveDir)
            if (past > overshootDegField) overshootDegField = past
        }

        // Settling time: first entry into the tolerance band after the target change.
        if (
            settleTimeSecField.isNaN() &&
                abs(errorRad) < SETTLE_POS_TOL_RAD &&
                abs(loadVelRad) < SETTLE_VEL_TOL_RAD
        ) {
            settleTimeSecField = tSec - moveStartTime
        }

        // Lost motion / lash gap: peak since the last reversal.
        if (lashGapDegField > lostMotionDegField) lostMotionDegField = lashGapDegField

        engagedEma += ENGAGED_EMA_ALPHA * ((if (engaged) 1.0 else 0.0) - engagedEma)
    }

    /** Clear all state (used on engine reset). */
    fun reset() {
        moveStartTime = 0.0
        moveDir = 0.0
        overshootDegField = 0.0
        settleTimeSecField = Double.NaN
        lostMotionDegField = 0.0
        targetRad = 0.0
        lastErrorDeg = 0.0
        lashGapDegField = 0.0
        engagedEma = 1.0
        sawTarget = false
    }

    fun steadyStateErrorDeg(): Double = lastErrorDeg

    fun overshootDeg(): Double = overshootDegField

    fun settleTimeSec(): Double = settleTimeSecField

    fun lashGapDeg(): Double = lashGapDegField

    fun lostMotionDeg(): Double = lostMotionDegField

    fun pctEngaged(): Double = engagedEma * 100.0

    companion object {
        private val SETTLE_POS_TOL_RAD = Math.toRadians(2.0)
        private val SETTLE_VEL_TOL_RAD = Math.toRadians(5.0)
        private const val ENGAGED_EMA_ALPHA = 0.02 // ~ last few seconds at 60 Hz
    }
}
